File: app/src/main/java/com/najmi/oreamnos/curator/IContentCurator.java
package com.najmi.oreamnos.curator;

import java.util.List;

/**
 * Interface defining content curation contract.
 * Allows swapping between different AI providers (Gemini, OpenAI, Groq, etc.)
 * without changing the calling code.
 */
public interface IContentCurator {

    /**
     * Curates the input text into a social media post.
     *
     * @param inputText     The text to curate
     * @param includeSource Whether to include source citation
     * @param keepStructure Whether to preserve original formatting/structure
     * @return The curated post
     * @throws Exception if curation fails
     */
    String curatePost(String inputText, boolean includeSource, boolean keepStructure) throws Exception;

    /**
     * Refines an existing post based on selected refinement options.
     *
     * @param originalPost  The post to refine
     * @param refinements   List of refinement options (e.g., "rephrase", "formal")
     * @param includeSource Whether to include source citation
     * @return The refined post
     * @throws Exception if refinement fails
     */
    String refinePost(String originalPost, List<String> refinements, boolean includeSource) throws Exception;

    /**
     * Gets the last prompt token count from API response.
     */
    int getLastPromptTokens();

    /**
     * Gets the last candidate (response) token count from API response.
     */
    int getLastCandidateTokens();

    /**
     * Gets the last total token count from API response.
     */
    int getLastTotalTokens();
}

File: app/src/main/java/com/najmi/oreamnos/curator/IContentCurator.kt
package com.najmi.oreamnos.curator

/**
 * Interface defining content curation contract.
 * Allows swapping between different AI providers (Gemini, OpenAI, Groq, etc.)
 * without changing the calling code.
 */
interface IContentCurator {

    /**
     * Curates the input text into a social media post.
     */
    @Throws(Exception::class)
    suspend fun curatePost(
        inputText: String,
        includeSource: Boolean,
        keepStructure: Boolean,
        length: String? = null
    ): String

    /**
     * Curates the input text with streaming token support.
     * Calls [onToken] for each token/segment as it arrives.
     * Returns the complete curated text when done.
     */
    @Throws(Exception::class)
    suspend fun curatePostStreaming(
        inputText: String,
        includeSource: Boolean,
        keepStructure: Boolean,
        length: String? = null,
        onToken: (String) -> Unit
    ): String

    /**
     * Sends [prompt] to the AI provider exactly as-is, with no additional system
     * prompt or post-processing. Used by [CardDataExtractor] to get raw JSON output
     * without the Malay social-media framing that [curatePost] applies.
     *
     * Implementations SHOULD override this. The default falls back to [curatePost]
     * for providers that have not yet implemented it.
     */
    @Throws(Exception::class)
    suspend fun generateRaw(prompt: String): String =
        curatePost(inputText = prompt, includeSource = false, keepStructure = true, length = null)

    /**
     * Refines an existing post based on selected refinement options.
     */
    @Throws(Exception::class)
    suspend fun refinePost(originalPost: String, refinements: List<String>, includeSource: Boolean): String

    val lastPromptTokens: Int
    val lastCandidateTokens: Int
    val lastTotalTokens: Int
}
